design-documents/frontend/admin-guidelines.md

```diff
@@ -0,0 +1,198 @@
+  - [React](#react)
+  - [State management](#state-management)
+  - [Code splitting](#code-splitting)
+  - [i18n](#i18n)
+  - [Theming](#theming)
+  - [GraphQL client](#graphql-client)
+  - [TypeScript import helpers](#typescript-import-helpers)
+- [devDependencies](#devdependencies)
+  - [Craco](#craco)
+  - [Redux mock store](#redux-mock-store)
+  - [Testing library](#testing-library)
+
+### version
+
+The version of this `package.json` follows [semver](https://semver.org/) and will change with this document accordingly.
+
+```json
+"version": "0.1.0",
+```
+
+### scripts
+
+Customize your CRA configuration with [craco][].
+
+- [Don't eject](https://create-react-app.dev/docs/alternatives-to-ejecting).
+- Avoid forking, because you can oftentimes do the same thing by creating your own [craco][] plugin. See [craco plugins on npm](https://www.npmjs.com/search?q=craco%20plugin) and the [craco recipes](https://github.com/sharegate/craco/tree/master/recipes) for guidance.
+
+```json
+"scripts": {
+  "start": "craco start",
+  "build": "craco build",
+  "test": "craco test"
+},
+```
+
+### [browserslist](https://github.com/browserslist/browserslist#readme)
+
+Don't forget your [polyfills](https://create-react-app.dev/docs/supported-browsers-features#supported-language-features)!
+
+```json
+"browserslist": [
+  ">0.2%",
+  "not dead",
+  "not ie < 11",
+  "not op_mini all"
+],
+```
+
+Refer also to [Magento DevDocs | Supported browsers](https://devdocs.magento.com/guides/v2.3/install-gde/system-requirements_browsers.html).
+
+### dependencies
+
+#### Utility functions
+
+Use vanilla (native) JS if possible, utility functions if not. [You don't (may not) need Lodash/Underscore](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#you-dont-may-not-need-lodashunderscore-). See also [just](https://github.com/angus-c/just#just).
+
+```json
+"@queso/camel-case": "^0",
+"@queso/kebab-case": "^1",
+```
+
+#### SSR
+
+Prepare for [SSR](https://medium.com/walmartlabs/the-benefits-of-server-side-rendering-over-client-side-rendering-5d07ff2cefe8) by using [universal fetching](https://www.npmjs.com/package/ky-universal). Don't use [axios](https://www.npmjs.com/package/axios) or [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) directly.
+
+```json
+"ky": "^0",
+"ky-universal": "^0",
+```
+
+#### React
+
+```json
+"react": "^16",
+"react-dom": "^16",
+"react-helmet": "^5",
+```
+
+#### Routing
+
+See [The Future of React Router and @reach/router](https://reacttraining.com/blog/reach-react-router-future/). Basically, the new API will be much closer to `@reach/router` than React Router, so this prepares us for that future.
+
+```json
+"@reach/router": "^1",
+```
+
+#### State management
+
+[Immutability Helper](https://www.npmjs.com/package/immutability-helper) was [rewritten in TypeScript](https://github.com/kolodny/immutability-helper/pull/123) and strikes a balance between features and size.
+
+```json
+"immutability-helper": "^3",
```

Comment:
[Immer](https://github.com/immerjs/immer) is another immutability library that's been gaining a lot of steam. It allows you to write all of your code in the normal way [].push(), mutating object properties (even nested ones), etc and then makes the changes immutably under the hood. The value of being able to just write normal mutations and not have to learn new APIs (especially one as ugly as immutability helper) is worth the extra 2.8 kb IMO. Check out the [benefits/usage section](https://github.com/immerjs/immer#benefits) for more examples. As a side note, Immer is the immutability library that the Redux team decided to ship with Redux Starter Kit.

Reply:
All good points. I've been holding onto this one for a while, but I think it's time to "let it go." Let me update this PR to recommend Immer. Thanks for challenging this point!

Reply:
Addressed in https://github.com/magento/architecture/pull/231/commits/b1c19bc24ff6a13764b7aa3d7e9a2c64daaebb16